nflow-engine/src/main/java/io/nflow/engine/workflow/definition/AbstractWorkflowDefinition.java

```diff
@@ -18,47 +19,102 @@
   }
 
-  protected AbstractWorkflowDefinition(String type, S initialState, S errorState, WorkflowSettings settings) {
+  /**
+   * Create a workflow definition with given settings and automatically scanned state methods.
+   *
+   * @param type
+   *          The unique identifier of this workflow definition.
+   * @param initialState
+   *          The default start state of the workflow. The state is automatically registered as one of the allowed states in this
+   *          workflow.
+   * @param errorState
+   *          The default error state of the workflow. The state is automatically registered as one of the allowed states in this
+   *          workflow.
+   * @param settings
+   *          The configuration for the workflow instances of this workflow type.
+   */
+  protected AbstractWorkflowDefinition(String type, WorkflowState initialState, WorkflowState errorState,
+      WorkflowSettings settings) {
     this(type, initialState, errorState, settings, null);
   }
 
-  protected AbstractWorkflowDefinition(String type, S initialState, S errorState, WorkflowSettings settings,
-      Map<String, WorkflowStateMethod> stateMethods) {
+  /**
+   * Create a workflow definition with given settings and state methods.
+   *
+   * @param type
+   *          The unique identifier of this workflow definition.
+   * @param initialState
+   *          The default start state of the workflow. The state is automatically registered as one of the allowed states in this
+   *          workflow.
+   * @param errorState
+   *          The default error state of the workflow. The state is automatically registered as one of the allowed states in this
+   *          workflow.
+   * @param settings
+   *          The configuration for the workflow instances of this workflow type.
+   * @param stateMethods
+   *          The state methods to be used for the states of this workflow type.
+   */
+  protected AbstractWorkflowDefinition(String type, WorkflowState initialState, WorkflowState errorState,
+      WorkflowSettings settings, Map<String, WorkflowStateMethod> stateMethods) {
     Assert.notNull(initialState, "initialState must not be null");
     Assert.isTrue(initialState.getType() == WorkflowStateType.start, "initialState must be a start state");
     Assert.notNull(errorState, "errorState must not be null");
     this.type = type;
     this.initialState = initialState;
     this.errorState = errorState;
     this.settings = settings;
+    WorkflowDefinitionScanner scanner = new WorkflowDefinitionScanner();
     if (stateMethods != null) {
       this.stateMethods = stateMethods;
     } else {
-      this.stateMethods = new WorkflowDefinitionScanner().getStateMethods(getClass());
+      this.stateMethods = scanner.getStateMethods(getClass());
     }
-    requireStateMethodExists(initialState);
+    registerState(initialState);
+    registerState(errorState);
+    scanner.getStaticWorkflowStates(getClass()).forEach(this::registerState);
```

Comment:
I thought the documentation said the states need to be registered. So do we need this scanner?
And if we do a scanner I would prefer that it would prefer a `static Collection<WorkflowState> getWorkflowStates()` method if such thing exists.
That would allow easier extending of workflow classes, where the extending class could filter out some parent states.
Maybe only if the method does not exist then we could register automatically?
Or can this be added later?